fawltydeps/packages.py

```diff
@@ -328,16 +336,57 @@ def temp_installed_requirements(requirements: List[str]) -> Iterator[Path]:
             (venv_dir / ".installed").touch()
             yield venv_dir
 
-    def lookup_packages(self, package_names: Set[str]) -> Dict[str, Package]:
+    # This function is only tagged as a context manager to facilitate an
+    # intechangeable use with `temp_installed_requirements()`
+    @staticmethod
+    @contextmanager
+    def cached_venv_with_requirements(
+        requirements: List[str], cache: pytest.Cache
+    ) -> Iterator[Path]:
+        """Install the given requirements into a cached venv.
+
+        Provide a path to the cached venv.
+        The venv persists after the call (and the installed deps are not
+        uninstalled).
+        The `--upgrade` and `--force-reinstall` options are not used to avoid
+        parallel tests over-writing the same package.
+        """
+
+        venv_dir = Path(
+            cache.mkdir(
+                f"fawltydeps_reused_venv_{sys.version_info.major}.{sys.version_info.minor}"
+            )
+        )
+        if not (venv_dir / ".installed").is_file():
+            venv.create(venv_dir, clear=True, with_pip=True)
+        subprocess.run(
+            [f"{venv_dir}/bin/pip", "install", "--no-deps"] + requirements,
+            check=True,  # fail if any of the commands fail
+        )
+        (venv_dir / ".installed").touch()
+        yield venv_dir
+
+    def lookup_packages(
+        self,
+        package_names: Set[str],
+    ) -> Dict[str, Package]:
         """Convert package names into Package objects via temporary pip install.
 
         Use the temp_installed_requirements() above to `pip install` the given
         package names into a temporary venv, and then use LocalPackageResolver
         on this venv to provide the Package objects that correspond to the
         package names.
         """
-        logger.info("Installing dependencies into a new temporary Python environment.")
-        with self.temp_installed_requirements(sorted(package_names)) as venv_dir:
+        if self.cache:
+            venv_context = self.cached_venv_with_requirements(
+                sorted(package_names), self.cache
+            )
+        else:
+            venv_context = self.temp_installed_requirements(sorted(package_names))
+            logger.info(
+                "Installing dependencies into a new temporary Python environment."
+            )
+        with venv_context as venv_dir:
```

Comment:
This code can stay more or less the same IMHO, but with the use of the `.cached_venv` class member and the two context managers above it looks like this in my version:

```python
from functools import partial

...

        if self.cached_venv is None:
            installed = self.temp_installed_requirements
            logger.info("Installing dependencies into a temporary Python environment.")
        else:
            installed = partial(self.installed_requirements, self.cached_venv)
            logger.info(f"Installing dependencies into {self.cached_venv}.")
        with installed(sorted(package_names)) as venv_dir:
```